### DISCUSSION TITLE
Buttons sequence as operation mode change trigger?

### DISCUSSION THREAD

**Discussion by chienMouille** [danielnilsson9/bbs-fw#196] [Q&A] [answered]
Hi,

Would it be possible to implement a more flexible configuration for choosing how the operation mode change is triggered?

For instance, I think it could be great to be able to do a sequence of 3 of the display buttons (like a PIN basically) to activate off road mode, and then a single press on menu button for instance to go back to standard mode.

(BBS02B here)

**Comment by danielnilsson9** [comment]
Not possible, this is only firmware for controller. Display does not send button presses to controller. Only sends selected assist level number, light output state, and walk assist state.

**Reply by chienMouille** [reply]
I see. So would it be possible then if the said PIN would be a sequence of assist levels state changes? Like one up one down one up one down for instance.  The menu button is not transferred to the controller I guess then. 

Thanks!

**Reply by danielnilsson9** [reply]
Yes I guess that could be possible to implement. It is  quite a bit of work to implement something fully configurable like that and I have previously said I won't do it. I think some people have modded the firmware themselves to have their own custom combination no one can know of.

**Reply by chienMouille** [reply]
Yes, I get it. Any chance you could point me to where I could implement/mod this? 
Also, any ressource on how to compile it then?

Thanks a lot

**Reply by chienMouille** [reply]
I guess another way would be to make a custom display, which just sends a light state change for instance after à the custom button combination. would need to implement the other commands too of course. 

Do you have a reference of the serial communication protocol you implemented ?

**Reply by danielnilsson9** [reply, marked as answer]
Current mode switch based on light switch is implemented here:
https://github.com/danielnilsson9/bbs-fw/blob/39c373b208a82aa0c3b66520c33543c8ef9835ba/src/firmware/app.c#L206

The other signal from display is just above it: app_set_assist_level()

Build instructions are on wiki:
https://github.com/danielnilsson9/bbs-fw/wiki/Build-from-Source

No reference on the display protocol, it is revered engineered from communication between original firmware and display, implemented here: https://github.com/danielnilsson9/bbs-fw/blob/master/src/firmware/extcom.c

If you were to develop a custom display a different protocol might be preferred, this one is very limited.

**Reply by chienMouille** [reply]
thank you so much!
I'll post here somewhere if I end up building a custom display!
